yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -253,10 +269,26 @@ private[yarn] class YarnAllocator(
    * added in recent versions.
    */
   private def createContainerRequest(resource: Resource): ContainerRequest = {
+    // filter nodes which don't have containers but node preference is required.
+    val nodes = {
+      val unusedNodes = preferredNodeLocationToUses.filterKeys(_ == false).keys
+      if (unusedNodes.isEmpty) {
+        null
+      } else {
+        unusedNodes.toArray
+      }
+    }
+
+    val racks = if (preferredRackLocations.isEmpty) {
+      null
+    } else {
+      preferredRackLocations.toArray
+    }
+
     nodeLabelConstructor.map { constructor =>
-      constructor.newInstance(resource, null, null, RM_REQUEST_PRIORITY, true: java.lang.Boolean,
+      constructor.newInstance(resource, nodes, racks, RM_REQUEST_PRIORITY, true: java.lang.Boolean,
```

Comment:
(btw, meant "without locality requirements", which is probably where the confusion crept in)